### ISSUE TITLE
extract_datetime fails when only month supplied

### ISSUE THREAD

**Issue by krisgesling** [MycroftAI/lingua-franca#20]
If an utterance like "when is september" is parsed using extract_datetime, it results in:
> ValueError: time data 'september' does not match format '%B %d %Y'

There doesn't seem to be an easy answer to what should be returned as this would be dependent on other contextual factors eg 
- On 29 August 2019 - "when is september" might be 1 September 2019.
- On 29 December 2019 - "when is september" might be 1 September 2020.
- On 29 December 2019 - "when was september" might be 30 September 2019.

So thought it was worth opening up the conversation on how this situation should be handled?

**Comment by JarbasAl** [comment]
i think a good default for the current method would be day 1 at 00:00

i propose we solve this issue by using this default and create two new utils, the real issue here is using the utils properly and not really a matter of functionality

- extract dates, list of any dates extracted
   - "yesterday was monday, tomorrow is wednesday" -> list with 2 dates
   - "every tuesday"  -> return a list of dates from anchorDate to Y

- extract date range, tuple of start / end date
   - "today" -> 00:00, 23:59
   - "next week" -> sunday 00:00, friday 23:59

**Comment by krisgesling** [comment]
I like the extra methods but I wonder if it would it be better for `extract_date("september")` to return None so it's clear that a specific date was not found?

**Comment by JarbasAl** [comment]
but there is a date, a date was indeed found, it's generic but still a date.

which use cases can you think off using only a month? in those cases what behaviour makes sense? day 1 hour 0 makes sense to me....

If you want to set a reminder for a once a year thing, a month only in the query would make sense, "remember me to X in september", returning None is not only unexpected in this case, but arguably plain wrong

**Comment by krisgesling** [comment]
My thinking is that if I'm checking for a date, and it's not found then I may check for a date range, whilst if it returns a successful date then I take the more specific case. 

However if I know that a month will return the first {month}, perhaps it just encourages checking for a date range first. This would then depend on how specific we are in the date range parsing, eg would "1 September" or "Friday next week" return None.

If I get a valid response from both `extract_date` and `extract_date_range`, then I have a difficult question of which one to use.

**Comment by JarbasAl** [comment]
it would depend a lot on the kind of skill / thing you are building, different tools for different problems, but both should notice the dates even if returning a different result, 

returning None should always mean "no date at all in this string", if the date is vague, it probably means the user does not need much detail, returning None is just confusing, this depends a lot on the specific application and developers need to design with this in mind

keeping voice assistants in mind there are two relevant use cases:

-  timers and reminders, you don't really care about date ranges, extracting a single date makes most sense here
- asking about a calendar, you will want to use a date range to check which events fall on it

**Comment by JarbasAl** [comment]
we could argue that returning a date range is better than returning a single date, since it basically contains the same information with an error interval

after implementing this the current extract_date can probably be refactored to just parse the output of date_range, keeping the relevant logic all in one place

**Comment by ChanceNCounter** [comment]
This seems like as good a place as any to chime in for posterity. I think `extract_datetime` should iterate over degrees of specificity until it gets a good match.

Take the alarm skill, since that's the skill with the failing test that got me thinking about it: extract_datetime craps out when it needs exact results that it isn't getting.

A skill like that might be easier to develop if the logic ran along the lines of, "Is that exact time in the collection I'm reading? How about the same hour? Is there anything on the same day? Two results. Which one is closer? That's not very close. speak_dialog('I found two results, please specify:')"

**Comment by retrodaredevil** [comment]
I'm currently making a skill where it makes sense to use an exact date, and sometimes a time range. It would be nice to have something like `extract_datetime_range` which would be something like this:
```
# I created this example in March btw
extract_datetime_range("this month") -> (datetime(2020, 3, 1), datetime(2020, 3, 31))
extract_datetime_range("last month") -> (datetime(2020, 2, 1), datetime(2020, 2, 29))
extract_datetime_range("in January") -> (datetime(2020, 1, 1), datetime(2020, 1, 31))
extract_datetime_range("last week") -> (dates from last week starting with Sunday to Saturday or whatever the language/local standard is (sometimes weeks start on Monday))
```

**Comment by ChanceNCounter** [comment]
@retrodaredevil

Makes sense. How do you envision using the output? Is a tuple the way?

**Comment by retrodaredevil** [comment]
Yeah I imagine a tuple is most intuitive. I'm going to be passing the output to a function that takes a start date (from date) and an end date (to date) which gives me solar energy information data. So a real world scenario would be like this:

"How much energy did I used last week?" -> Data from last week starting Sunday to Saturday
"How much energy did I use last year?" -> Data from last year starting January 1 to December 31
"How much energy did I use this month?" -> Data from start of this month to current day

For something like "this month", I think the extract_datetime_range should also return a future date, but for what I'm doing, I won't be using any future dates as data for them doesn't exist. 

I also came across something when just using the current implementation of `extract_datetime`: If I ask for December 1, it gives me December 1, 2020. In my case December 1, 2019 is what I wanted because I was asking for past data. It would be nice to have an optional argument in `extract_datetime` called something like `prefer_past_date`.

**Comment by ChanceNCounter** [comment]
Indeed. It's not as intuitive, but I have no idea how long it might be until a fix gets to prod, so ysk:

extract_datetime() takes an anchor date as a parameter. It currently tries to output the *next* instance of an ambiguous date, because that will *usually* be the desired result. For example, in Mycroft, "Set an appointment for December 1st."

So, as a temporary workaround, if you will *always* be looking at *last year*, set the anchor date to datetime.now().year. If not, have your skill pass a specific year to that param.

I hate to suggest that you write code we'll render obsolete Soon, but who knows if it *is* Soon?

**Comment by ChanceNCounter** [comment]
In yet another episode of Function Signatures Get Huge, we'd need the user to be able to specify when the week starts.

**Comment by retrodaredevil** [comment]
I was thinking that should default to the language/region being used, but could also be a setting in preferences. `extract_datetime`'s arguments aren't too long yet.

**Comment by JarbasAl** [comment]
> In yet another episode of Function Signatures Get Huge, we'd need the user to be able to specify when the week starts.

i'm ok with huge signature functions, in the end each language is effectively pre configured to just work, and the user retains power to customize. Don't be afraid of keyword arguments !

We should be opinionated regarding strong defaults, but not introduce artificial limitations. This is above all for developers, let's allow them to get wild while ensuring casual users (skill makers) can also use it easily and out of the box.

**Comment by JarbasAl** [comment]
#96 closes this
